include/alpaka/block/shared/st/BlockSharedMemStMember.hpp

```diff
@@ -0,0 +1,146 @@
+                        /*virtual*/ ~BlockSharedMemStMemberImpl() = default;
+
+                        template <typename T>
+                        void alloc() const
+                        {
+                            uint8_t* buf = &m_mem[m_allocdBytes];
+                            new (buf) T();
+                            m_allocdBytes += alignPitch<T>();
+#if (defined ALPAKA_DEBUG_OFFLOAD_ASSUME_HOST) && (! defined NDEBUG)
+                            ALPAKA_ASSERT(m_allocdBytes < m_capacity);
+#endif
+                        }
+
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align" // "cast from 'unsigned char*' to 'unsigned int*' increases required alignment of target type"
+#endif
+                        template <typename T>
+                        T& getLatestVar() const
+                        {
+                           return *reinterpret_cast<T*>(&m_mem[m_allocdBytes-alignPitch<T>()]);
+                        }
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic pop
+#endif
+
+                        void free() const
+                        {
+                            m_allocdBytes = 0u;
+                        }
+
+                    };
+                }
+                //#############################################################################
+                //! Static block shared memory provider using a pointer to
+                //! externally allocated fixed-size memory, likely provided by
+                //! BlockSharedMemDynMember.
+                template<unsigned int TDataAlignBytes = 4>
+                class BlockSharedMemStMember :
+                    public detail::BlockSharedMemStMemberImpl<TDataAlignBytes>,
+                    public concepts::Implements<ConceptBlockSharedSt, BlockSharedMemStMember<TDataAlignBytes>>
+                {
+                public:
+                    using detail::BlockSharedMemStMemberImpl<TDataAlignBytes>::BlockSharedMemStMemberImpl;
+                };
+
+                namespace traits
+                {
+                    //#############################################################################
+                    template<
+                        typename T,
+                        unsigned int TDataAlignBytes,
+                        std::size_t TuniqueId>
+                    struct AllocVar<
+                        T,
+                        TuniqueId,
+                        BlockSharedMemStMember<TDataAlignBytes>>
+                    {
+                        //-----------------------------------------------------------------------------
+                        static auto allocVar(
+                            block::shared::st::BlockSharedMemStMember<TDataAlignBytes> const &smem)
+                        -> T &
+                        {
+                            smem.template alloc<T>();
+                            return smem.template getLatestVar<T>();
```

Comment:
Is there a reason for the split here? Can't `alloc` directly return the variable as well?

Reply:
Yes. This is in order for facilitate derived classes which need to synchronize to do the `alloc()`.